client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java

```diff
@@ -119,14 +133,49 @@ public RestClientBuilder setRequestConfigCallback(RequestConfigCallback requestC
     }
 
     /**
+     * Sets the base path for every request used by the http client.
+     * <p>
+     * In essence, every request's {@code endpoint} is prefixed by this {@code basePath}. The base path is useful for when Elasticsearch is
+     * behind a proxy that provides a base path; it is not intended for other purposes.
+     *
+     * @throws NullPointerException if {@code basePath} is {@code null}.
+     * @throws IllegalArgumentException if {@code basePath} is empty, only '/', or ends with more than one '/'.
+     */
+    public RestClientBuilder setBasePath(String basePath) {
+        Objects.requireNonNull(basePath, "basePath must not be null");
+        String cleanBasePath = basePath.trim();
+
+        if (cleanBasePath.startsWith("/") == false) {
+            cleanBasePath = "/" + cleanBasePath;
+        }
+
+        // best effort to ensure that it looks like "/base/path" rather than "/base/path/"
+        if (cleanBasePath.endsWith("/")) {
+            cleanBasePath = cleanBasePath.substring(0, cleanBasePath.length() - 1);
+
+            if (cleanBasePath.endsWith("/")) {
+                throw new IllegalArgumentException("basePath is malformed. too many trailing slashes: [" + basePath + "]");
+            }
+        }
+
+        if (cleanBasePath.isEmpty() || "/".equals(cleanBasePath)) {
+            // just don't set it
```

Comment:
I left that simple comment there incase a user hit the exception and searched our code. I'll remove it though because I added the Javadoc afterward.